File: C++/Algorithms/Searching/Binary_Search.cpp
// Title : Binary Search Implementation Using C++
// Author : Ansh Walia <anshwalia@outlook.com>
// Date Created : 18/07/2020

#include <iostream>
#include <stdlib.h>
#include <cmath>

using namespace std;

// Function to sort array ( Bubble Sort )
void sort(int * array,int size){
	// Temporary Variable used for swapping
	int temp = 0;

	// Sorting Loop
	for(int i = 0; i < (size-1); i++){
		for(int j = (i+1); j < size; j++){
			if(array[j] < array[i]){
				temp = array[i];
				array[i] = array[j];
				array[j] = temp;
			}
		}
	}
}

// Function to display/print array
void displayArray(int array[],int size){
	cout << "[ ";
	for(int i = 0; i < size; i++){
		cout << array[i] << ", ";
	}
	cout << "]\n";
}

// Binary Search Algorithm
int binarySearch(int array[],int lowerBound,int upperBound,int value){

	if(upperBound >= lowerBound){
		int mid = ( lowerBound + round((upperBound-lowerBound)/2));

		// If value present at mid
		if(value == array[mid]){
			return mid;
		}

		// If value is smaller than value at mid
		if(value < array[mid]){
			return binarySearch(array,lowerBound,(mid-1),value);
		}
		// If value if greater than value at mid
		else{
			return binarySearch(array,(mid+1),upperBound,value);
		}
	}
	else{
		return -1;
	}

}

int main(){
	// Variable to store array size
	int size = 0;

	// Variable to store value to search
	int value = 0;

	// Array Size Input
	cout << "Enter Array Size : ";
	cin >> size;

	// Dynamiclly Allocating Memory for Array
	int * array = (int*)malloc( size * sizeof(int) );

	// Array Data Input
	for(int i = 0; i < size; i++){
		cout << "Element at Index [" << i << "] : ";
		cin >> array[i];
	}

	// Value to search input
	cout << "\nEnter Value to Search : ";
	cin >> value;

	// Before Sorting
	cout << "\nArray Before Sorting : ";
	displayArray(array,size);

	// Sorting Array Elements
	sort(array,size);

	// After Sorting
	cout << "\nArray After Sorting : ";
	displayArray(array,size);

	// Searching
	int match = binarySearch(array,0,(size-1),value);

	// If match is not equal to -1 result is printed
	if(match != -1){
		cout << "\nMatch Found => Index : " << match << " , Value : " << array[match] << "\n\n";
	}
	else{
		cout << "\nNo Match Found!\n\n";
	}

	// Freeing/Clearing Allocated Memory
	free(array);

	return 0;
}

File: JavaScript/Algorithms/Searching/Binary_Search.js
// Title : Binary Search Implementation using JavaScript
// Author : Ansh Walia <anshwalia@outlook.com>
// Date Created : 20/07/2020

'use strict';

// Binary Search
function binarySearch(array,lowerBound,upperBound,searchValue){
    
    if(upperBound >= lowerBound){
        // Calculating Array Mid-Point
        let midPoint = (lowerBound + Math.floor((upperBound-lowerBound)/2));

        // Evaluating if value at midpoint is equal to search value
        if(searchValue == array[midPoint]){
            return midPoint;
        }

        // Evaluating if value at midpoint is greater than search value
        if(searchValue < array[midPoint]){
            return binarySearch(array,lowerBound,(midPoint-1),searchValue);
        }
        // Evaluating if value at midepoint is less than search value
        else{
            return binarySearch(array,(midPoint+1),upperBound,searchValue);
        }

    }
    else{
        // Returning -1 if no matching value found
        return -1;
    }

}

// Bubble Sort : For Sorting Array
function bubbleSort(array){
    // Temporary variable to help in swapping
    let temp = 0;

    // Sorting Loop
    for(let i = 0; i < (array.length-1); i++){
        for(let j = (i+1); j < array.length; j++){
            if(array[j] < array[i]){
                // Element Swapping
                temp = array[i];
                array[i] = array[j];
                array[j] = temp;
            }
        }
    }

    // Returning Sorted Loop
    return array;
}

// Test Function
async function test(){
    // Test Array
    let testArr = [10,9,8,7,6,1,2,3,4,5];

    // Value to search in test array
    let searchValue = 8;

    // Printing array before sorting
    console.log("\nArray Before Sorting :",testArr,"\n");

    // Calling bubbleSort function to sort test array 
    testArr = await bubbleSort(testArr);

    // Printing array after sorting
    console.log("Array After Sorting :",testArr,"\n");

    // Calling binarySearch function to search given value in array
    let matchIndex = await binarySearch(testArr,0,(testArr.length-1),searchValue);

    // If match index value is less than 0 i.e -ve than element is not present in give array
    if(matchIndex < 0){
        console.log("No Match Found!\n");
    }
    // if match index value is greater than or equalt to 0 than element is present in given array
    else{
        console.log("Match found at index :",matchIndex,"\n");
    }
}

// Start Point
test();